ark/jsonschema/object.ts

```diff
@@ -0,0 +1,265 @@
+			ctx.reject({
+				expected: `an object JSON Schema with at most ${jsonSchema.maxProperties} required properties`,
+				actual: `an object JSON Schema with ${jsonSchema.required!.length} required properties`
+			})
+		}
+
+		const jsonSchemaObjectMaxPropertiesValidator = (
+			data: object,
+			ctx: Traversal
+		) => {
+			const keys = Object.keys(data)
+			return keys.length <= maxProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at most ${maxProperties} propert${maxProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${maxProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMaxPropertiesValidator)
+	}
+	if ("minProperties" in jsonSchema) {
+		const minProperties = jsonSchema.minProperties
+
+		const jsonSchemaObjectMinPropertiesValidator = (
+			data: object,
+			ctx: Traversal
+		) => {
+			const keys = Object.keys(data)
+			return keys.length >= minProperties ?
+					true
+				:	ctx.reject({
+						expected: `an object with at least ${minProperties} propert${minProperties === 1 ? "y" : "ies"}`,
+						actual: `an object with ${keys.length.toString()} propert${minProperties === 1 ? "y" : "ies"}`
+					})
+		}
+		predicates.push(jsonSchemaObjectMinPropertiesValidator)
+	}
+	return predicates
+}
+
+const parsePatternProperties = (
+	jsonSchema: JsonSchema.Object,
+	ctx: Traversal
+) => {
+	if (!("patternProperties" in jsonSchema)) return
+
+	const patternProperties = Object.entries(jsonSchema.patternProperties).map(
+		([key, value]) => [new RegExp(key), parseJsonSchema(value)] as const
+	)
+
+	// Ensure that the schema for any property is compatible with any corresponding patternProperties
+	patternProperties.forEach(([pattern, parsedPatternPropertySchema]) => {
+		Object.entries(jsonSchema.properties ?? {}).forEach(
+			([property, schemaForProperty]) => {
+				if (!pattern.test(property)) return
+
+				const parsedPropertySchema = parseJsonSchema(schemaForProperty)
+
+				if (!parsedPropertySchema.overlaps(parsedPatternPropertySchema)) {
+					ctx.reject({
+						path: [property],
+						expected: `a JSON Schema that overlaps with the schema for patternProperty ${pattern} (${parsedPatternPropertySchema.description})`,
+						actual: parsedPropertySchema.description
+					})
+				}
+			}
+		)
+	})
+
+	// NB: We don't validate compatability of schemas for overlapping patternProperties
+	// since getting the intersection of regexes is inherently non-trivial.
+	const indexSchemas = patternProperties.map(
+		([pattern, parsedPatternPropertySchema]) => ({
+			signature: { domain: "string" as const, pattern: [pattern] },
+			value: parsedPatternPropertySchema.internal
+		})
+	)
+	return indexSchemas
+}
+
+const parsePropertyNames = (jsonSchema: JsonSchema.Object, ctx: Traversal) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = parseJsonSchema(jsonSchema.propertyNames)
+
+	if (
```

Comment:
Seems like you'd want to be checking `propertyNamesValidator.extends(type.string)` here

Reply:
Yes, although if I do that then L111 errors since we're no longer checking the `domain` is present.

I'm not sure what to include in the error message instead of the domain.

Reply:
I went with `.toString()`: e1842780